Refactor to integrate pkg ticketbuyer for automated  ticket purchases

tuxcanfly opened the pull request:
This PR integrates the newly refactor pkg `ticketbuyer` in dcrticketbuyer to make automated ticket purchases.

https://github.com/decred/dcrticketbuyer/pull/70

Refs #387

jrick commented on config.go:
```diff
@@ -110,6 +130,25 @@ type config struct {
 	ProxyUser        string `long:"proxyuser" description:"Username for proxy server"`
 	ProxyPass        string `long:"proxypass" default-mask:"-" description:"Password for proxy server"`
 
+	// Automatic ticket buyer settings
```
I think all of these settings can be removed, as long as we're ok with using a client to enable the ticket purchasing with the specific parameters.  For "easy" (parameterless) ticket buying these aren't needed.

tuxcanfly replied:
OK, removed. Passing defaults for all the configuration options instead. Config can be updated later, with provided specific parameters, if necessary.

jrick commented on config.go:
```diff
@@ -573,6 +632,11 @@ func loadConfig() (*config, []string, error) {
 		}
 	}
 
+	// Always prompt passphrase if stake mining is enabled
```
Instead of modifying the config, it just leave the config as is and when cfg.EnableStakeMining is true it can just run the password prompt at startup anyways.

tuxcanfly replied:
OK, updated.

jrick commented on dcrwallet.go:
```diff
@@ -137,6 +137,9 @@ func walletMain() error {
 		}
 		startWalletRPCServices(w, rpcs, legacyRPCServer)
+		if w.StakeMiningEnabled {
+			go startTicketPurchase(w)
```
I think this will need some additional parameters.

We need the private passphrase because ticket purchasing requires an unlocked wallet.  When not running the wallet with stake mining already enabled (--enablestakemining), it should still be possible to turn on (and off!) the ticket purchaser.  When used from the grpc interface, the passphrase will need to be passed in as a parameter.

It should also pass in some sort of default options here for the "easy" --enablestakemining option.  When started via RPC, more advanced parameters can be used instead of the defaults.

tuxcanfly replied:
OK, started passing in the default options here.

For toggling ticket purchaser, I think we could have rpc similar to `setgenerate`. Will be adding that.

As for passphrase, would it make sense to let the caller make sure the wallet is unlocked before turning on purchasing? If it's not unlocked, the toggle purchase rpc call can return an error.

jrick replied:
well in the gRPC server there is no notion of an unlocked wallet. the private passphrase is passed as a field in every RPC request message that requires it.  And we want this usable over grpc

tuxcanfly replied:
Updated to use passphrase if any and unlock the wallet.

jrick commented on log.go:
```diff
@@ -13,6 +13,7 @@ import (
 	"github.com/btcsuite/seelog"
 
 	"github.com/decred/dcrrpcclient"
+	"github.com/decred/dcrticketbuyer/ticketbuyer"
```
ticketbuyer should be a dcrwallet package, not part of the dcrticketbuyer repo.  When dcrticketbuyer is updated to use the wallet's built in ticket purchasing, it can simply be an rpc client that interacts with the wallet's ticket purchaser over RPC.

tuxcanfly replied:
Done. The pkg `ticketbuyer` is now part of wallet.

jrick commented on ticketbuyer.go:
```diff
@@ -0,0 +1,108 @@
+		PriceTarget:        cfg.PriceTarget,
+		TicketAddress:      cfg.TicketAddress,
+		TxFee:              cfg.TicketFee,
+	}, dcrdClient, dcrwClient, activeNet.Params)
```
I guess this should be obvious from the rest of my comments, but when this package is moved to this repo, it should interact with the wallet directly using Go apis, and not use any RPC.

tuxcanfly replied:
Thanks, this has been updated. The package is now a part of this repo and the caller can pass in wallet functions, so now it is not dependent on RPC.

jrick commented on dcrwallet.go:
```diff
@@ -137,6 +138,30 @@ func walletMain() error {
 		startWalletRPCServices(w, rpcs, legacyRPCServer)
+		if w.StakeMiningEnabled {
+			ticketbuyerCfg := &ticketbuyer.Config{
+				AccountName:        "default",
```
Think this should use cfg.PurchaseAccount

tuxcanfly replied:
Yup.

tuxcanfly commented on ticketbuyer.go:
```diff
@@ -0,0 +1,93 @@
+// startTicketPurchase launches ticketbuyer to start purchasing tickets.
+func startTicketPurchase(w *wallet.Wallet, ticketbuyerCfg *ticketbuyer.Config) {
+	retryDuration := time.Second * 5
+	for {
```
Agreed, will do that.

tuxcanfly replied:
Done. Caller needs to pass ChainClient, moved the caller to an appropriate location in `dcrwallet.go`

jrick commented on ticketbuyer.go:
```diff
@@ -0,0 +1,93 @@
+		tkbyLog.Debugf("Retrying chain client connection in %v", retryDuration)
+	}
+	dcrdClient := w.ChainClient().Client
+	walletCfg := &ticketbuyer.WalletCfg{
```
From my initial perusal through this code, this type looks unneeded.  We have access to the wallet so we don't need an extra level of indirection to call methods from it.  Looks like rpc leftovers.

tuxcanfly replied:
This is required to make sure the pkg `ticketbuyer` works both standalone (with dcrwallet rpc client) as well as being used here. For example, here's the caller in `dcrticketbuyer`: 

https://github.com/tuxcanfly/dcrticketbuyer/blob/pkg-ticketbuyer/main.go#L251

jrick replied:
oh, I see.  So dcrticketbuyer is still doing all the purchasing logic then.  Makes sense because there's no rpc interface yet to start the purchaser or change its parameters.  Once that is in I think this can be removed.

jrick replied:
actually I think we could go ahead and clean this up here now.  dcrticketbuyer can be left alone for now (keeping its own old purchasing logic) and only after the RPC interfaces are added will we want to update it to use those.

tuxcanfly replied:
OK, working on that. I guess it would involve some code duplication but that should be fixed when we phase out the purchase logic in dcrticketbuyer.

jrick replied:
yeah, we can afford some code duplication during the transition, that's no problem.

tuxcanfly replied:
Done, using the wallet directly in `ticketbuyer` now.

jrick commented on ticketbuyer/example_test.go:
```diff
@@ -0,0 +1,5 @@
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package ticketbuyer_test
```
if there's nothing here let's remove the file.

tuxcanfly replied:
Removed.

jrick commented on wallet/chainntfns.go:
```diff
@@ -38,6 +38,22 @@ func (w *Wallet) handleConsensusRPCNotifications(chainClient *chain.RPCClient) {
 				return w.onBlockConnected(tx, n.BlockHeader, n.Transactions)
 			})
+
+			if w.StakeMiningEnabled && w.purchaser != nil {
```
read of w.purchaser is not safe for concurrent access.

jrick replied:
I think this also shows that the current design is backwards.  wallet should not call the ticket purchaser.  it should emit an event that causes the started ticket purchaser goroutine (if any) to begin purchasing tickets.

tuxcanfly replied:
Done. Using `TransactionNotifications` to trigger purchases now.

tuxcanfly commented on dcrwallet.go:
```diff
@@ -137,6 +138,30 @@ func walletMain() error {
+				PoolAddress:        "",
+				MaxFee:             1.0,
+				MinFee:             0.01,
+				FeeSource:          "mean",
```
Updated to use consts for defaults.

jrick commented on ticketbuyer/mempool.go:
```diff
@@ -0,0 +1,96 @@
+package ticketbuyer
```
please add a copyright header

tuxcanfly replied:
OK, added.

jrick commented on ticketbuyer/price.go:
```diff
@@ -0,0 +1,126 @@
+// when the chain has not yet sufficiently synced to the chain tip.
+var vwapHeightOffsetErrStr = "beyond blockchain tip height"
+
+// containsHeightOffsetError specifies whether or not the reponse
```
missing rest of comment

tuxcanfly replied:
Updated. Looks like it was missing from the first commit.

jrick commented on ticketbuyer.go:
```diff
@@ -0,0 +1,101 @@
+
+// purchaseManager is the main handler of websocket notifications to
+// pass to the purchaser and internal quit notifications.
+type purchaseManager struct {
```
I would really like to see this not part of the main package. Any reason in particular it was added here?

tuxcanfly replied:
Done, moved to the pkg. Also passing in wallet, to make purchasing dynamic based on `setgenerate`.

jrick commented on ticketbuyer.go:
```diff
@@ -0,0 +1,101 @@
+		case v := <-p.ntfnChan:
+			if v != nil {
+				for _, block := range v.AttachedBlocks {
+					go p.purchase(int64(block.Height))
```
I think that it should purchase tickets whenever the main chain tip changes, and not several times for each new block in the main chain after a reorganize.

tuxcanfly replied:
I think dcrticketbuyer was getting block notifications from dcrd, so it might have been only purchasing at the tip as you mentioned. Will cross check that.

tuxcanfly replied:
Leaving it for now, since I see dcrticketbuyer was making purchases for every block received in the notification.

jrick commented on dcrwallet.go:
```diff
@@ -189,9 +190,10 @@ func walletMain() error {
+// flag is set.
 func startPromptPass(w *wallet.Wallet) {
-	promptPass := cfg.PromptPass
+	promptPass := cfg.PromptPass || cfg.EnableStakeMining
```
on second thought this is probably a bad idea since it will break scripts.  Probably want to just continue requiring the user to manually unlock the wallet without timeout.

tuxcanfly replied:
OK, restored as before. Purchasing requires unlocked wallet, manually.

tuxcanfly commented:
Thanks, updated as per the above suggestions. Please review.

Also added the ability to read `ticketbuyer.conf` from appdata, so that users migrating from `dcrticketbuyer` can just copy their old conf file. Some of the options from the old conf are superseded by newer options in wallet conf, added deprecation warnings for them. Since this is only intended for migrating users, we're not parsing the command line flags for these new options.

jrick commented on ticketbuyer.go:
```diff
@@ -0,0 +1,233 @@
+
+	// A config file in the current directory takes precedence.
+	exists := false
+	if _, err := os.Stat(defaultTicketBuyerConfigFilename); !os.IsNotExist(err) {
```
I'd remove all this. We don't want stuff in the current directory ever taking precedence.  Also doesn't seem to any effect on behavior since if it does exist, it just sets the config file option to the same as the default.

If the config file is missing just return the default config.

```
if _, err := os.Stat(fileName); os.IsNotExist(err) {
    return cfg, nil
}
```

I also think this should occur *after* the main program's config parsing, so if an alternate appdata dir was set, the config will be looked there and not using the default appdata setting.  Please pass that value in as a parameter here and make the default relative to that directory.

tuxcanfly replied:
Thanks, will do that.

tuxcanfly replied:
OK, using path relative to appdata and returning default if missing.

jrick commented on ticketbuyer.go:
```diff
@@ -0,0 +1,233 @@
+
+	// Check deprecated aliases.
+	if cfg.AccountName != defaultAccountName {
+		fmt.Fprintln(os.Stderr, "accountname option has been replaced by "+
```
Since this parsing should be done after the main program config's parsing, and logging should already be set up and ready to use, print these warnings to the log.

jrick commented on ticketbuyer.go:
```diff
@@ -0,0 +1,233 @@
+			"wallet option ticketfee -- please update your config")
+	}
+
+	// Create the home directory if it doesn't already exist.
```
definitely not needed. if the files exists, the directory it is in already exists too

tuxcanfly replied:
OK, removed.

jrick commented on ticketbuyer/purchase.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+// NotificationHalder handles notifications, which trigger ticket purchases.
+func (p *PurchaseManager) NotificationHalder() {
```
this is misspelled

tuxcanfly replied:
Thanks, fixed.

jrick commented on glide.lock:
```diff
@@ -1,12 +1,14 @@
   version: 73889fb79bd687870312b6e40effcecffbd57d30
 - name: github.com/btcsuite/fastsha256
   version: 637e656429416087660c84436a2a035d69d54e2e
+- name: github.com/btcsuite/go-flags
```
I don't see addition of this to the glide.yaml.. but anyways this shouldn't be added.  We switched to upstream repo pkg path.  You can see `github.com/jessevdk/go-flags` further down in the lockfile.  Please change all uses of btcsuite/go-flags to jessevdk/go-flags to avoid pulling the same package twice under different names.

tuxcanfly replied:
Thanks, done.

tuxcanfly commented on dcrwallet.go:
```diff
@@ -285,6 +290,36 @@ func rpcClientConnectLoop(legacyRPCServer *legacyrpc.Server, loader *wallet.Load
+			tcfg, err := loadTicketBuyerConfig()
+			if err != nil {
+				log.Errorf("Unable to load ticket buyer config: %v", err)
+				return
```
OK, done.

jrick commented on dcrwallet.go:
```diff
@@ -285,6 +286,36 @@ func rpcClientConnectLoop(legacyRPCServer *legacyrpc.Server, loader *wallet.Load
 			}
+			tcfg, err := loadTicketBuyerConfig(cfg.AppDataDir)
+			if err != nil {
+				log.Errorf("Unable to load ticket buyer config: %v", err)
```
Can you also add to this message that it won't purchase tickets?

tuxcanfly replied:
Thanks, added.

jrick commented on dcrwallet.go:
```diff
@@ -289,6 +289,7 @@ func rpcClientConnectLoop(legacyRPCServer *legacyrpc.Server, loader *wallet.Load
 			tcfg, err := loadTicketBuyerConfig(cfg.AppDataDir)
 			if err != nil {
 				log.Errorf("Unable to load ticket buyer config: %v", err)
+				log.Info("Ticket purchase is disabled")
```
purchasing

dajohi commented:
Tested OK with 2 wallets sharing a seed.
